### PR TITLE
records: return new upsert result + retry upsert

### PR DESCRIPTION
**Pull request by TBonnin**
## Describe your changes

This PR contains 2 commits:
- intermediary step before removing the legacy records logic: still double writing but now returning the result of the upsert to new db
- retrying the entire upsert transaction didn't solve the deadlock_detected error. I am still not sure what's going on, making an attempt at only retrying the upsert query inside the transaction

### CODE REVIEW

File: packages/records/lib/models/records.ts [hunk 1]
```diff
@@ -433,3 +428,11 @@ async function getRecordsByExternalIds(external_ids: string[], connection_id: nu
 
     return result;
 }
+
+function shouldRetryUpsert(error: Error): boolean {
+    if ('code' in error) {
+        const errorCode = (error as { code: string }).code;
+        return errorCode === '40P01'; // deadlock_detected
```
Comment (khaliqgant) [inline packages/records/lib/models/records.ts]: nit: seems a bit random to have this here

File: packages/persist/lib/controllers/persist.controller.ts [hunk 2]
```diff
@@ -117,17 +110,10 @@ class PersistController {
...
-                .then((res) => {
-                    if (isErr(res)) {
-                        throw res.err;
-                    }
-                })
-                .catch((err: unknown) => {
-                    logger.error(`Failed to delete records: ${stringifyError(err)}`);
-                });
-            return await dataService.upsert(legacyRecords, nangoConnectionId, model, activityLogId, environmentId, true, logCtx);
+            const newUpsert = recordsService.upsert(records, nangoConnectionId, model, true);
```
Comment (khaliqgant) [inline packages/persist/lib/controllers/persist.controller.ts:113]: Unclear what the `true` is doing/signaling here. Feel like this is left over from code I probably wrote...

**Review by khaliqgant — APPROVED** [review]
Non blocking nits

File: packages/persist/lib/controllers/persist.controller.ts (hunk 2 above)
Comment (TBonnin) [reply packages/persist/lib/controllers/persist.controller.ts:113]: it is `softDelete`. Passing a struct with named fields would be better. I will make a PR

File: packages/records/lib/models/records.ts (hunk 1 above)
Comment (TBonnin) [reply packages/records/lib/models/records.ts]: I had it inlined in `upsert` originally. I am happy to put it back there if you prefer

Comment (khaliqgant) [reply packages/records/lib/models/records.ts]: If there will be other error types that we should retry on then I would leave it here. If this is a one off then would inline it. Either way is fine to me tbh

Comment (TBonnin) [reply packages/records/lib/models/records.ts]: back to being inlined. thx

File: packages/persist/lib/controllers/persist.controller.ts (hunk 2 above)
Comment (TBonnin) [reply packages/persist/lib/controllers/persist.controller.ts:113]: addressed in https://github.com/NangoHQ/nango/pull/2080

**Pull request merged**